spock-core/src/main/java/org/spockframework/mock/runtime/IMockMaker.java

```diff
@@ -0,0 +1,317 @@
+  interface IMockMakerSettings {
+
+    /**
+     * Creates a simple {@link IMockMakerSettings} object with only an ID.
+     *
+     * @param id the mock maker ID.
+     * @return the settings object
+     */
+    static IMockMakerSettings simple(String id) {
+      return new IMockMakerSettings() {
+        @Override
+        public String getMockMakerId() {
+          return id;
+        }
+
+        @Override
+        public String toString() {
+          return id;
+        }
+      };
+    }
+
+    String getMockMakerId();
+  }
+
+  /**
+   * {@link MockMakerCapability} describes the possible features of an {@link IMockMaker}.
+   *
+   * <p>This provides the ability to extend the list of capabilities in the future without breaking existing external {@link IMockMaker}s.</p>
+   *
+   * @since 2.4
+   */
+  @Beta
+  enum MockMakerCapability {
+    /**
+     * Is able to mock an interface type.
+     */
+    INTERFACE("Cannot mock interfaces."),
+    /**
+     * Is able to mock a class type.
+     */
+    CLASS("Cannot mock classes."),
+    /**
+     * Is able to mock a class/interface type with additional interfaces.
+     */
+    ADDITIONAL_INTERFACES("Cannot mock classes with additional interfaces."),
+    /**
+     * Is able to pass explicit constructor arguments to the class constructor.
+     */
+    EXPLICIT_CONSTRUCTOR_ARGUMENTS("Explicit constructor arguments are not supported."),
+    /**
+     * Is able to mock a final class type.
+     */
+    FINAL_CLASS("Cannot mock final classes."),
+    /**
+     * Is able to mock a final method.
+     */
+    FINAL_METHOD("Cannot mock final methods."),
+    /**
+     * Is able to mock a static method.
+     *
+     * <p>Note: This is currently not supported by spock-core, this is for future extension.</p>
+     */
+    STATIC_METHOD("Cannot mock static methods."),
+    ;
+    private final IMockabilityResult notMockable;
+
+    MockMakerCapability(String errorMessage) {
+      this.notMockable = () -> errorMessage;
+    }
+
+    /**
+     * @return IMockabilityResult with the default error message if this {@link MockMakerCapability} is not present.
+     */
+    IMockabilityResult notMockable() {
+      return notMockable;
+    }
+  }
+
+  /**
+   * {@link IStaticMock} is returns by {@link IMockMaker#makeStaticMock(IMockCreationSettings)}, which allows
+   * {@link IMockMaker} to create static mocks, if supported.
+   *
+   * <p>
+   * The caller need to enable the static mock with {@link #enable()}, before usage. And need to {@link #disable()} after usage.
+   *
+   * <p>
+   * The implemented static mock requires to be implemented as thread-local behavior, so {@link #enable()} does only affect the current thread.
+   *
+   * @since 2.4
+   */
+  @Beta
+  interface IStaticMock {
+
+    /**
+     * @return the static mocked type
+     */
+    Class<?> getType();
+
+    /**
+     * Enables the static mock.
+     * Note: The default state on creation is disabled.
+     */
+    void enable();
```

Comment:
This is very low level API, where normally the call to enable and disable are at different places, so I do not think the API with `AutoCloseable` would help. Especially no user should ever need to call these low-level primitives.
Only Spock-internals need to call that, when a static mock is created, and then the `disable` is registered for the cleanup phase.

Reply:
Oh, I see.
Then I misunderstood it due to the partial implementation.
I somehow thought that in the end the user has to enable and disable the static mock manually, something like:
```groovy
def foo() {
  def staticPatternMock = Mock(Pattern, static: true)
  
  when:
    staticPatternMock.enable().withCloseable {
      sut.doSomethingThatCallsStaticMethod()
    }

  then:
    1 * staticPatternMock.compile(_)
}
```
If it's only for spock internal usage at creation and cleanup time it of course makes not too much sense.

Reply:
You can see an example for my API proposal in the Draft for static support #1756.

```
given:
MockStatic(StaticClass)
when:
def result = StaticClass.staticMethod()
then:
1 * StaticClass.staticMethod() >> MOCK_VALUE
result == MOCK_VALUE
```

Reply:
As a side note: The API with `def staticPatternMock = Mock(Pattern, static: true)` 
would be really cumbersome to implement and use:

1. We can't create a mock for some final classes, like classes of the JRE, so we can't return an instance to that.
2. The `Mock` method would pass `Class<Mock>` into the closure as parameter, which is broken in the CodeCompletion and on Groovy calling site.
3. Some named args in the options do not make sense in the static case, like `constructorArgs`

So I would prefer to define new methods for the static case, which would not return a mock instance and handle the `Closure` correctly.

Reply:
> You can see an example for my API proposal in the Draft for static support

Ah, I see. Yeah, that looks much better then and quite spocky. :-)

I don't intend to do a full discussion here now, as it belongs at best to that other draft PR.
Just some comments in return to your comments, but please let's keep it with that for this PR: :-)

> We can't create a mock for some final classes, like classes of the JRE

Well, that is not really true. It actually depends on the effort you put in and / or the MockMakers you integrate.
You can mock almost any class, latest when you register an instrumentation agent that does some of the work.
For certain edge cases it will not work when you attach the agent at runtime, then you have to manually configure the agent in the JVM settings of the test process. Latest with that you can change all classes, maybe except 2 or so that are even then loaded too early like `String`. And if you are desperate enough, you can even modify those last cases by doing some up-front offline instrumentation and prepending them to the bootclasspath.

If you now say that all this is just theoretic bull****, no it is even reality already. Sarek already provides all these capabilities. It just waited for this PR to maybe then get a seamless integration with Spock: https://github.com/SarekTest/Sarek

So the "we can't" actually only refers to "while using mockito-inline only". :-)

> Some named args in the options do not make sense in the static case, like constructorArgs

That's no reason at all actually.
Not with the current state, as that is already the sad truth.
For example try to combine `constructorArgs` with `useObjenesis` and the latter is ignored (should throw though imho as mentioned earlier).
Or try to combine `constructorArgs` with `mockMaker: MockMakers.javaProxy`.
Or try to create a `Mock(...)` with `global: true`.
...

> So I would prefer to define new methods for the static case, which would not return a mock instance and handle the Closure correctly.

I'm pretty sure this will become discussed at length in that other PR. :-D